### ISSUE TITLE
Feat: Container runtime for functions

### ISSUE THREAD

**Issue by rayli09** [sst/sst#4462]
## Context
V2 supports container runtime for [functions](https://docs.sst.dev/constructs/Function#configuring-container-runtime), which was crucial to monorepo setup with multiple micro-services in different language runtimes. However we're considering migrating to ion but found container runtime isn't supported yet.

We do see that ion supports all pulumi providers, does that mean we can use pulumi's aws provider to setup lambda functions with container runtime as a temporary workaround?

**Comment by lefnire** [comment]
Has anyone gotten Pulumi example above working? I get this error using it via `npm run dev`

![Image](https://github.com/user-attachments/assets/7a124208-2ab9-43b6-9dd9-4fc9a843659e)

And I'm assuming it's because SST manages what gets copied from source directories into .sst/platform/*. And since SST isn't aware of this raw Pulumi code, it doesn't copy the assets over. I've tried manually copying the Dockerfile but no cigar. I'm assuming then this Pulumi container function should be managed separately (not using sst deploy), and any needed ARNs should be copy/pasted?

**Comment by aplotor** [comment]
@lefnire i was able to deploy a similar example to above. i ran into the `no such file or directory` err too. for the image `context` path, it's relative to `.sst/platform/`, so prefix the path with `../../` or whatever you need to reach the dockerfile dir

**Comment by danielsharvey** [comment]
I'm keen for this.

My workaround for using containers for Lambda is to piggy-back off the python feature:

```ts
const mysqlHelper = new sst.aws.Function('mysqlHelper', {
  handler: 'index.handler',
  bundle: 'mysql-helper-docker',
  hook: {
    postbuild: async (dir) => {
      const buildDir = join($cli.paths.work, 'artifacts', 'mysqlHelper-src');
      execSync(`cp -a ${dir}/* ${buildDir}/`);
    },
  },
  runtime: 'nodejs20.x',
  python: { 
    container: true,
  },
  copyFiles: [
    { from: 'mysql-helper-docker', to: '.' },
  ],
  // ^ this is not executed???
  memory: '10240 MB',
  storage: '1024 MB',
  timeout: '15 minutes',
  ...
});
```

Note that when you use `bundle` the `copyFiles` files are not copied (I think that is the discriminator).

**Comment by truejoolean** [comment]
> Has anyone gotten Pulumi example above working? I get this error using it via `npm run dev`
> 
> ![Image](https://github.com/user-attachments/assets/7a124208-2ab9-43b6-9dd9-4fc9a843659e)
> 
> And I'm assuming it's because SST manages what gets copied from source directories into .sst/platform/*. And since SST isn't aware of this raw Pulumi code, it doesn't copy the assets over. I've tried manually copying the Dockerfile but no cigar. I'm assuming then this Pulumi container function should be managed separately (not using sst deploy), and any needed ARNs should be copy/pasted?

I am also getting this error. Updated the context as suggested by @aplotor , but still running into that issue

`context: "../../src/containers/my-container-example",`

Has anyone been able to solve the "b.Va is not a function" error? I've tried juggling around with different pulumi versions, but no luck so far.

```
"@pulumi/aws": "^6.42.0",
"@pulumi/awsx": "^2.17.0",
"sst": "^3.17.0",
```

**Comment by jdwitten** [comment]
I've been following this issue because our team is using container functions with V2 and have been hoping to upgrade to V3 with that feature supported. Curious if there's a timeline for this?

Also a bit confused by some of the examples in the [docs](https://sst.dev/docs/examples/#aws-lambda-python-container) that show container functions are supported for Python? Is this true for v3? and why is only Python supported in Docker container? Shouldn't supporting container functions be agnostic of what is actually running inside of the container?

**Comment by 55Cancri** [comment]
> I've been following this issue because our team is using container functions with V2 and have been hoping to upgrade to V3 with that feature supported. Curious if there's a timeline for this?
> 
> Also a bit confused by some of the examples in the [docs](https://sst.dev/docs/examples/#aws-lambda-python-container) that show container functions are supported for Python? Is this true for v3? and why is only Python supported in Docker container? Shouldn't supporting container functions be agnostic of what is actually running inside of the container?

Yes I saw that too, what an odd choice. Why only python? In my case, I want to use a nodejs container so I can ship my own version of playwright chromium browsers and not have to rely on @sparticuz/chromium. @jayair is supporting nodejs container still planned?

**Comment by danielsharvey** [comment]
Just confirming: I am still using my [workaround](https://github.com/sst/sst/issues/4462#issuecomment-2766413892) detailed above.

TLDR; Node runtime + python container:true + postbuild to copy files.

**Comment by EthanShoeDev** [comment]
Thought I might try nodejs v22 or deno to get bytecode caching for my large ssr lambda. Tmk this is not very feasible with zip deployments because aws could update nodejs runtime to a patch version and it would invalidate bytecode generated at build time.

**Comment by tommyOtsai** [comment]
Hi @fwang , Wondering if there are any plans to support container runtime for lambda functions outside of Python on SST v3?
